crates/rpc/rpc/src/eth/cache/mod.rs

```diff
@@ -331,11 +369,54 @@ where
                                     // looking up the tree is blocking
                                     let res = provider
                                         .find_block_by_hash(block_hash, BlockSource::Database);
+
                                     let _ = action_tx
                                         .send(CacheAction::BlockResult { block_hash, res });
                                 }));
                             }
                         }
+                        CacheAction::GetBlockWithSenders { block_hash, response_tx } => {
+                            if let Some(block) = this.full_block_cache.get(&block_hash).cloned() {
+                                let _ = response_tx.send(Ok(Some(block)));
+                                continue
+                            }
+
+                            // block is not in the cache, request it if this is the first consumer
+                            if this
+                                .full_block_cache
+                                .queue(block_hash, ResponseSender::BlockWithSenders(response_tx))
+                            {
+                                let provider = this.provider.clone();
+                                let action_tx = this.action_tx.clone();
+                                let rate_limiter = this.rate_limiter.clone();
+                                this.action_task_spawner.spawn_blocking(Box::pin(async move {
+                                    // Acquire permit
+                                    let _permit = rate_limiter.acquire().await;
+                                    // Only look in the database to prevent situations where we
+                                    // looking up the tree is blocking
+                                    let res = provider
+                                        .find_block_by_hash(block_hash, BlockSource::Database);
```

Comment:
this step is redundant we only do this to get the block number.

we should add a new helper function to the provider that takes the hash instead:

`block_with_senders_by_hash`:

https://github.com/paradigmxyz/reth/blob/main/crates/storage/provider/src/traits/block.rs#L115-L124